### PR TITLE
RFC: rich retrieval results

### PR DESCRIPTION
**Pull request by baskaryan**
Could be more opinionated about DocumentResult and RetrievalResult attributes if we wanted

Related to #4710

### CODE REVIEW

File: libs/core/langchain_core/retrievers.py [hunk 1]
```diff
@@ -22,8 +24,19 @@
         Callbacks,
     )
 
+
+class DocumentResult(Serializable):
+    document: Document
+    metadata: dict = Field(default_factory=dict)
+
+
+class RetrievalResult(Serializable):
```
Comment (eyurtsev) [inline libs/core/langchain_core/retrievers.py:33]: This API works IMO

File: libs/core/langchain_core/retrievers.py [hunk 2]
```diff
@@ -22,8 +24,19 @@
         Callbacks,
     )
 
+
```
Comment (eyurtsev) [inline libs/core/langchain_core/retrievers.py:27]: Do we have examples of the metadata that folks want to add?

I would consider potentially adding some standard metadata the root level to force an identical schema between providers

```python
class DocumentResult(Serializable):
    document: Document
    extra: dict = Field(default_factory=dict)
    id: Optional[str]
    score: Optional[float]   
```

Comment (eyurtsev) [reply libs/core/langchain_core/retrievers.py:27]: In many search engines, what's returned is not the Document, but a hit snippet. A snippet is a chunk of text from the original document that matched.


```python
class Hit(Serializable):
    source: Optional[Document] # <-- The actual source of the document 
    snippet: ... # <-- Some type we don't know yet that might represent a chunk of the document
    extra: dict = Field(default_factory=dict)
    id: Optional[str]
    score: Optional[float]   
```

File: libs/core/langchain_core/retrievers.py (hunk 1 above)
Comment (baskaryan) [reply libs/core/langchain_core/retrievers.py:33]: any argument to be made that we should explicitly include `query: str`?

**Pull request closed without merging**